app/src/main/java/org/schabi/newpipe/util/StreamItemAdapter.java

```diff
@@ -271,13 +289,147 @@ public static <X extends Stream> Single<Boolean> fetchSizeForWrapper(
+                    final MediaFormat format = MediaFormat.getFromSuffix(suffix);
+                    if (format != null) {
+                        streamsWrapper.setFormat(stream, format);
+                        return true;
+                    }
+                }
+            }
+            return false;
+        }
+
+        @VisibleForTesting
+        public static <X extends Stream> boolean retrieveMediaFormatFromContentDispositionHeader(
+                @NonNull final X stream,
+                @NonNull final StreamInfoWrapper<X> streamsWrapper,
+                @NonNull final Response response) {
+            // parse the Content-Disposition header,
+            // see https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Disposition
+            // there can be two filename directives
+            String contentDisposition = response.getHeader("Content-Disposition");
+            if (contentDisposition != null) {
+                try {
+                    contentDisposition = Utils.decodeUrlUtf8(contentDisposition);
+                    final String[] parts = contentDisposition.split(";");
+                    for (String part : parts) {
+                        final String fileName;
+                        part = part.trim();
+                        // extract the filename
+                        if (part.startsWith("filename=")) {
+                            // remove directive and decode
+                            fileName = Utils.decodeUrlUtf8(part.substring(9));
+                        } else if (part.startsWith("filename*=")) {
+                            // The filename* directive passes the encoding of the filename, too.
+                            // The encoding is prepended to the actual filename
+                            // and the file name *should* be put into quotes.
+                            // There needs to be a character (mostly one of the quotes)
+                            // which separates the encoding from the filename
+                            part = part.substring(10); // remove directive
+                            final String encoding = Parser.matchGroup1(
+                                    "([0-9|A-Z|a-z|\\-|_]+)", part);
+                            // remove encoding and separating character
+                            part = part.substring(encoding.length() + 1);
+                            final Charset charset = Charset.forName(encoding);
+                            // decode info with the given encoding
+                            if (charset == null) {
+                                continue;
+                            }
+                            if (charset.equals(Charset.defaultCharset())) {
+                                fileName = part;
+                            } else {
+                                fileName = new String(part.getBytes(charset),
+                                        StandardCharsets.UTF_8);
```

Comment:
I am not sure whether this code does the right thing. If the data is already in a String then it has already been decoded in the default format, so the only way to re-decode it in a different charset would be to reverse the initial decoding (with `encode(default charset)`) and then decode it with the different charset.

Reply:
Hm, you are right. In that case, we cannot do much about it. And the encoding should not matter because we only want to get the file extension which should always be in from the Latin alphabet. Shall I remove the encoding part?

Reply:
Yes in my opinion

Reply:
Done